### PR TITLE
Refactor DLWP wrapper and add batching support

### LINKED ISSUE #137: 🚀[FEA]: Batched support for the DLWP wrapper/plugin
### Is this a new feature, an improvement, or a change to existing functionality?

New Feature

### How would you describe the priority of this feature request

Critical (currently preventing usage)

### Please provide a clear description of problem you would like to solve.

Currently the MIP plugin of DLWP only supports a single input. Batched support would be needed for better/faster inferencing workflows. 

### Describe any alternatives you have considered

_No response_

### PR DESCRIPTION
**Pull request by ktangsali**
# Modulus Pull Request

## Description
This PR adds batching support for the DLWP wrapper needed for faster inference and also updates the time stepping such that the time passed to the wrapper is now corresponds to the latest timestep input to the model. 

Closes #137 

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [x] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: modulus/models/fcn_mip_plugin.py [hunk 1]
```diff
@@ -159,21 +159,25 @@ def __init__(
...
         j = self.input_map_wts.col.values - 1
         data = self.input_map_wts.S.values
-        M = torch.sparse_coo_tensor(np.array((i, j)), data).to(device).type(dtype)
-
-        for i in range(len(input_list)):
-            input_list[i] = input_list[i].reshape(num_chans, -1) @ M.T
-            input_list[i] = input_list[i].reshape(1, num_chans, 6, 64, 64)
+        M = torch.sparse_coo_tensor(np.array((i, j)), data).type(dtype).to(device)
 
+        input = input.reshape(input.shape[0], input.shape[1], input.shape[2], -1) @ M.T
```
Comment (dallasfoster) [inline modulus/models/fcn_mip_plugin.py:168]: I found a bug while using Pytorch 1.13.1 that we should consider:

```python 
input.reshape(input.shape[0], input.shape[1], input.shape[2], -1) @ M.T
```

gives the error

```
NotImplementedError: Could not run 'aten::sparse_dim' with arguments from the 'CUDA' backend
```

while

```python
temp = input_list[i].reshape(num_ens * num_chans, -1)
temp = (M @ temp.T).T
```

does not. This is likely due to some limitation in matrix batch multiplication for sparse tensors in Pytorch, Consider revising here and in `prepare_output`.

File: modulus/models/fcn_mip_plugin.py [hunk 2]
```diff
@@ -159,21 +159,25 @@ def __init__(
...
+        input = input.reshape(input.shape[0], input.shape[1], input.shape[2], 6, 64, 64)
```
Comment (dallasfoster) [inline modulus/models/fcn_mip_plugin.py:169]: Can we abstract the size of the underlying grid?

File: modulus/models/fcn_mip_plugin.py [hunk 3]
```diff
@@ -205,25 +210,20 @@ def prepare_input(self, input, time):
...
         data = self.output_map_wts.S.values
-        M = torch.sparse_coo_tensor(np.array((i, j)), data).to(device).type(dtype)
+        M = torch.sparse_coo_tensor(np.array((i, j)), data).type(dtype).to(device)
 
-        for i in range(len(output_list)):
-            output_list[i] = output_list[i].reshape(num_chans, -1) @ M.T
-            output_list[i] = output_list[i].reshape(1, num_chans, 721, 1440)
-        output = torch.stack(output_list, dim=1)
+        output = output.reshape(output.shape[0], 2, output.shape[2], -1) @ M.T
+        output = output.reshape(output.shape[0], 2, output.shape[2], 721, 1440)
```
Comment (dallasfoster) [inline modulus/models/fcn_mip_plugin.py:226]: Again, abstract grid size.

**Review by dallasfoster — COMMENTED** [review]
Tested, looks good other than comments below.

File: modulus/models/fcn_mip_plugin.py (hunk 1 above)
Comment (ktangsali) [reply modulus/models/fcn_mip_plugin.py]: Great comment. I added a fix that just reshapes the data to a 2d array and then changes the shape again after multiplication. Can you have a look again?

File: modulus/models/fcn_mip_plugin.py (hunk 2 above)
Comment (ktangsali) [reply modulus/models/fcn_mip_plugin.py]: Great comment, done!

**Comment by ktangsali** [comment]
> Tested, looks good other than comments below.

Thanks @dallasfoster for the review. I have addressed most of the comments. Can you please have a look again?

File: modulus/models/fcn_mip_plugin.py [hunk 4]
```diff
@@ -165,8 +167,10 @@ def prepare_input(self, input, time):
         data = self.input_map_wts.S.values
         M = torch.sparse_coo_tensor(np.array((i, j)), data).type(dtype).to(device)
 
-        input = input.reshape(input.shape[0], input.shape[1], input.shape[2], -1) @ M.T
-        input = input.reshape(input.shape[0], input.shape[1], input.shape[2], 6, 64, 64)
+        N, T, C = input.shape[0], input.shape[1], input.shape[2]
+        input = input.reshape(N * T * C, -1) @ M.T
+        S = int((M.shape[0] / 6) ** 0.5)
```
Comment (dallasfoster) [inline modulus/models/fcn_mip_plugin.py:172]: So this is a fine choice but maybe why can't we have the size of the grid be defined in the xarray dataset file `self.input_map_wts` and `self.output_map_wts`?

**Review by dallasfoster — CHANGES_REQUESTED** [review]

File: modulus/models/fcn_mip_plugin.py [hunk 5]
```diff
@@ -165,8 +167,10 @@ def prepare_input(self, input, time):
         data = self.input_map_wts.S.values
         M = torch.sparse_coo_tensor(np.array((i, j)), data).type(dtype).to(device)
 
-        input = input.reshape(input.shape[0], input.shape[1], input.shape[2], -1) @ M.T
-        input = input.reshape(input.shape[0], input.shape[1], input.shape[2], 6, 64, 64)
+        N, T, C = input.shape[0], input.shape[1], input.shape[2]
+        input = input.reshape(N * T * C, -1) @ M.T
```
[the remark this answers is not in the document]
Comment (ktangsali) [reply modulus/models/fcn_mip_plugin.py]: Thanks @dallasfoster . There seems to be some weird pytorch mismatch that might be causing this as the CI passes. I will have to look into it deeper, but I agree that your proposed method should be much safer. I added it, please take another look.

File: modulus/models/fcn_mip_plugin.py (hunk 4 above)
Comment (ktangsali) [reply modulus/models/fcn_mip_plugin.py:172]: Yeah, let's retain the current way for now, but once we have some conventions on the map files, we can revisit this.

**Review by dallasfoster — APPROVED** [review]

**Review by NickGeneva — APPROVED** [review]

**Pull request merged**